### PR TITLE
state maintenance in DataParallel

### LINKED ISSUE #565: State maintenance in DP
In many image generation tasks with GANs, generator and discriminator is trained through the same generated image single iteration.
In PyTorch Lightning, the procedure is written like below:
```python
def training_step(self, batch, batch_nb, optimizer_i):
    foo = batch['foo']
    bar = batch['bar']

    if optimizer_i == 0:  # train discriminator
        self.foo_out = self.netG(foo)  # register as a instance variable

        # calc d_loss
        d_loss = ...

        return {'loss': d_loss}

    elif optimizer_i == 1:  # train generator
        # common reconstruction error
        g_loss = F.l1_loss(self.foo_out, bar)
        # other losses
        ...

        return {'loss': g_loss}
```
It works well on single GPU, however, `self.foo_out` has been flushed in `optimizer_i == 1` branch when DP is set.

I think it is a undesired behavior, any help or fix?

### PR DESCRIPTION
**Pull request by ananyahjha93**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
WIP. This fails on a few cases of variable declaration.
Tries to fix #565. 
#1649 is slightly different from simple state maintenance in **DataParallel.**

## PR review    
Need to be reviewed by core maintainers as it changes the way DataParallel works in default PyTorch.

## Did you have fun?
Yes! :)

### CODE REVIEW

**Comment by ananyahjha93** [comment]
@Borda hey, this is done from my side but there is merge conflict because I added another parent class to LightningModule.

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Added `DistributedStateHooks` class with methods `on_after_model_replicate()` and `on_after_dp_parallel_apply()`
```
Comment (Borda) [inline CHANGELOG.md:11]: link this PR
```suggestion
- Added `DistributedStateHooks` class with methods `on_after_model_replicate()` and `on_after_dp_parallel_apply()` ([#1895](https://github.com/PyTorchLightning/pytorch-lightning/pull/1895))
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -26,6 +28,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Changed
 
+- Allow users to maintain state in `DataParallel` mode using `self.distributed_state` container within `LightningModule`; solves ([#565](https://github.com/PyTorchLightning/pytorch-lightning/issues/565))
```
Comment (Borda) [inline CHANGELOG.md:31]: link this PR
```suggestion
- Allow users to maintain state in `DataParallel` mode using `self.distributed_state` container within `LightningModule` ([#1895](https://github.com/PyTorchLightning/pytorch-lightning/pull/1895))
```

File: pl_examples/basic_examples/state_maintenance_dp_demo.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            tqdm_dict = {'g_loss': g_loss}
+            output = OrderedDict({
+                'loss': g_loss,
+                'progress_bar': tqdm_dict,
+                'log': tqdm_dict
+            })
+            return output
+
+        # train discriminator
+        if optimizer_idx == 1:
```
Comment (Borda) [inline pl_examples/basic_examples/state_maintenance_dp_demo.py:142]:
```suggestion
        elif optimizer_idx == 1:
```
or it can be changed in the previous block?

**Review by justusschock — DISMISSED** [review]

**Comment by Borda** [comment]
@ananyahjha93 how about this one, mind rebase and resolve conflicts? 🐰

**Comment by marrrcin** [comment]
Is this going to be re-opened? https://github.com/PyTorchLightning/pytorch-lightning/issues/4353#issuecomment-720790786

**Comment by ananyahjha93** [comment]
@marrrcin Yes, I think this is required to solve the DP metric bug in #4353.

**Comment by Borda** [comment]
@ananyahjha93 can we get this done by end of this week to pass it in 1.1? 🐰

**Comment by Borda** [comment]
@ananyahjha93 how is going here? cc: @tchaton

**Comment by tchaton** [comment]
Hey @ananthsub,

Any updates on this one ?

Best,
T.C

**Comment by ananyahjha93** [comment]
@tchaton needs to be redone after accelerator refactor.

**Comment by vlievin** [comment]
Hi @ananyahjha93, any update on that one ?
